### ISSUE TITLE
Performance issues with large data sets

### ISSUE THREAD

**Issue by mwilliamson-healx** [graphql-python/graphene#268]
For our use case, we send a few thousand objects to the client. We're currently using a normal JSON API, but are considering using GraphQL instead. However, when returning a few thousand objects, the overhead of resolving values makes it impractical to use. For instance, the example below returns 10000 objects with an ID field, and that takes around ten seconds to run.

Is there a recommended way to improve the performance? The approach I've used successfully so far is to use the existing parser to parse the query, and then generate the response by creating dictionaries directly, which avoids the overhead of resolving/completing on every single value.

```python
import graphene

class UserQuery(graphene.ObjectType):
    id = graphene.Int()

class Query(graphene.ObjectType):
    users = graphene.Field(UserQuery.List())

    def resolve_users(self, args, info):
        return users

class User(object):
    def __init__(self, id):
        self.id = id

users = [User(index) for index in range(0, 10000)]

schema = graphene.Schema(query=Query)

print(schema.execute('{ users { id } }').data)
```

**Comment by ekampf** [comment]
@mwilliamson-healx Ive had the same problem.
Fortunately the `next` version of Graphene fixes the issue (and also adds performance tests to make sure it doesnt regress).
Though its a risk using the library's bleeding edge, Ive been running the `next` version (`pip install graphene>=1.0.dev`) for a couple of weeks now in production without problems.

So you should give it a try and see if it solves your problem (and if not, maybe there's some new performance test cases to add to Graphene's performance tests)

**Comment by syrusakbary** [comment]
@mwilliamson-healx as Eran pointed, the `next` version its been rewritten with a special focus on performance.

We also added a benchmark for a similar case you are exposing (retrieving about 100k elements instead of 10k).
https://github.com/graphql-python/graphene/blob/next/graphene/types/tests/test_query.py#L129

The time spent for retrieving 10k elements should be about 10-20 times faster now (50-100ms?).
https://travis-ci.org/graphql-python/graphene/jobs/156652274#L373

Please test in the `next` branch and let me know if you run into any non-performant case, I will happily work on that.

**Comment by mwilliamson-healx** [comment]
Thanks for the suggestion! I gave Graphene 1.0.dev0 a go, and while it's certainly faster, it still takes around a second to run the example above. Admittedly, I didn't try it out on the speediest of machines, but suggests that it would still be the dominant factor in response time for our real data.

**Comment by syrusakbary** [comment]
@mwilliamson-healx some of the performance bottleneck was also in the `OrderedDict` generation.
For that `graphql-core` uses `cyordereddict` when available (a implementation of `OrderedDict` in Cython that runs about 2-6x faster).

Could you try installing cyordereddict with `pip install cyordereddict` and running again the tests? (no need to modify anything in the code).

Thanks!

**Comment by syrusakbary** [comment]
PS: There are some plans to port some code to `Cython` to make `graphene`/`graphql-core` even more performant, however any other suggestion would be always welcome! :)

**Comment by mwilliamson-healx** [comment]
Thanks again for the suggestion! Using cyordereddict shaves about 200ms off the time (from 1s to 0.8s), so an improvement, but still not ideal. I had a look around the code, but nothing stuck out to me as an easy way of improving performance. The problem (from my extremely quick and poorly informed glance!) is that you end up resolving every single value, which includes going through any middleware and having to coordinate promises. Keeping that functionality while being competitive with just spitting out dicts directly seems rather tricky.

[The proof of concept I've got](https://github.com/mwilliamson/python-graphjoiner) sidesteps the issue somewhat by parsing the GraphQL query, and then relying on the object types being able to generate the requested data directly, without having to further resolve values. It's very much a proof of concept (so doesn't support fragments, and isn't really GraphQL compliant yet), but feel free to have a look. Assuming the approach is sane, then it's hard to see how to reconcile that approach with the normal GraphQL resolve approach.

**Comment by syrusakbary** [comment]
Hi @mwilliamson-healx,
At first congrats for your great proof of concept!

I've been thinking for a while how we can improve performance in GraphQL. This repository -`graphene`- uses `graphql-core` under the hood which is a very similar port of the GraphQL-js reference implementation.

The problem we are seeing is that either in `graphql-core` and `graphql-js` that **each type/value is checked in runtime** (*what I mean is that the resolution+serialization function is "discovered" in runtime each time a value is completed*). In js the performance difference is minimal as usually have a great `JIT` that optimizes each of the type/value completion calls. However as Python doesn't have any `JIT` by default, this result in a quite expensive operation.

In the current `graphql-js` and `graphql-core` implementations if you want to execute a GraphQL query this is how the process will look like:

```
Parse AST from string (==> validate the AST in the given schema) ==> Execute a AST given a Root type.
```

However we can **create a "_Query Builder_"** as intermediate step before executing that will know exactly what are the fields we are requesting and therefore it's **associated types** and **resolvers**, so we don't need to "search" for them each time we are completing the value.
This way, the process will be something like:

```
Parse AST from string (==> validate the AST in the given schema) ==> Build the Query resolver based in the AST ==> Execute the Query resolver builder given a Root type.
```

Your [proof of concept](https://github.com/mwilliamson/python-graphjoiner) is doing the latter so the performance difference is considerable comparing with the current `graphql-core` implementation.

I think it's completely reasonable to introduce this extra `Query resolver` build step before executing for avoid the performance bottleneck of doing it in runtime. In fact, I would love to have it in `graphql-core`.

And I also think this would be a great idea to add in the `graphql-js` implementation as it will improve performance there too and push other language implementations ( @leebyron ).

**Comment by mwilliamson** [comment]
Thanks for the kind words. One question I had was how much you'd imagine trusting the query builder? For my implementation, I was planning on putting the responsibility of correctness onto the queries (rather than having the GraphQL implementation check). The result is that, unlike the normal implementations of GraphQL, it's possible to implement something that doesn't conform to the GraphQL spec.

**Comment by syrusakbary** [comment]
I'm working in the query builder concept. As of right now the benchmarks shows about 4x improvement when returning large datasets.

Related PR in `graphql-core`: https://github.com/graphql-python/graphql-core/pull/74

**Comment by syrusakbary** [comment]
Some updates!
I've been working non-stop on keep improving the performance with the `Query Builder`.

# Benchmarks

## Retrieving 10k ObjectTypes
Doing something similar to the following query where `allContainers` type is a `[ObjectType]` and `x` is a `Integer`:
```graphql
{
  allContainers {
    x
  }
}
```
  
  - With Query Builder: https://travis-ci.org/graphql-python/graphql-core/jobs/158369656#L488
> 30ms

  - Without Query Builder: https://travis-ci.org/graphql-python/graphql-core/jobs/158369656#L484
> 350ms


## Retrieving 10k items in a List
Doing something similar to the following query where `allInts` type is a `[Integer]`
```graphql
{
  allInts
}
```
  
  - With Query Builder: https://travis-ci.org/graphql-python/graphql-core/jobs/158369656#L483
> 12ms

  - Without Query Builder: https://travis-ci.org/graphql-python/graphql-core/jobs/158369656#L486
> 30ms

# Conclusion

The work I'm doing so far is being a demonstration the code performance still have margins to improve while preserving fully compatibility with GraphQL syntax.

The proof of concept speedup goes between 3x and 10x while maintaining the syntax and features `GraphQL` have. Still a lot of work to do there, but it's a first approach that let us discover new paths for speed improvement.


# Extra

I think by using `Cython` for some critical instructions we can gain about another 10-20x in speed.


# Transport

Apart of using Cython I'm thinking how we can plug multiple kind of transports into GraphQL.
So instead of creating Python Objects each time we are accessing a field, and then transforming the result to `JSON`, another approach could be transform the values directly into `JSON` or whatever transport we are using.

This way the result could be created directly in the output format. This way we can plug other transports like `binary` (CapN Proto/FlatBuffers/Thrift/others), `msgpack` or any other thing we could think of.

**Comment by mwilliamson** [comment]
Thanks for working on this. I've taken a look at the proof of concept you wrote, but it's not clear to me exactly how it behaves, and how it's saving time versus the existing implementation. It seems like it's still resolving all fields of objects in the response, but I could easily have misread.

I adjusted my proof of concept to (optionally) integrate with GraphQL properly. This means that you can do things like generating the schema, introspect, and the all other stuff that GraphQL does, but it means you hit the performance penalty again. It seems to me that the easiest way of fixing this for my use case would be a way to prevent resolution from descending into the object that my proof of concept produces -- a way of returning a value from resolve functions that doesn't trigger resolution on any fields (since they're already resolved).

Perhaps something like:

```python
def resolve_users(...):
    ...
    return FullyResolvedValue(users)
```

where `users` is already fully resolved by inspecting the AST or whatever. Alternatively, a decorator on the function itself might be clearer.

This shifts more responsibility onto the calling code to make sure that the returned value is of the correct shape in order to ensure it's still a valid GraphQL implementation, but that's definitely a good trade-off for me.

**Comment by qubitron** [comment]
@syrusakbary any update on this thread? I am using graphene in production and unfortunately it simply doesn't scale for even the moderate data sets being returned by my API. I'm slowly rewriting my API calls as normal HTTP calls and seeing 10x RPS increases (and therefore 10x reduction in server costs), but it means I'm not losing the flexibility of the graphQL approach. Seems like the solution discussed in this thread would save me from this headache!

**Comment by mwilliamson-healx** [comment]
In case it's useful, I've been using the project I mentioned above in production, and performance has been good enough. In particular, it avoids having to run a (potentially asynchronous) resolver for every field. I'm still tweaking the API, but it should be reasonably stable (and better documented!) soon.

https://github.com/healx/python-graphjoiner

**Comment by syrusakbary** [comment]
Hi @qubitron,

I created the experimental branch `features/next-query-builder` in `graphql-core` with a new execution system that improves significantly the speed: https://github.com/graphql-python/graphql-core/pull/74/.

# How to use it

1. Install it with `pip install https://github.com/graphql-python/graphql-core/archive/features/next-query-builder.zip`

2. Enable the new executor
```python
from graphql.execution import executor

executor.use_experimental_executor = True
```

3. Execute the query

If you can try it and output here your results would be great!

# Extra questions

*To help us optimize for your use case*:
* Are you in a CPython environment? (non pypy or google app engine)
* Did you use any GraphQL middleware?

**Comment by qubitron** [comment]
@syrusakbary it took me a bit of time to get to a place where I had a good test for this. The package you provided seems to make a big improvement! Cutting total execution time for my request roughly in half, with the graphene portion reduced by a factor of 3x. 

Initially it wasn't working because I already had graphql-core installed, doing "pip uninstall graphql-core" before running your command above finally yielded the performance improvements.

More about my workload... I'm using a flask web server with graphene_sqlalchemy and returning objects that inherit from SQLAlchemyObjectType (not sure if that counts as middleware but I get similar results when I return plain graphene.ObjectType). 

For this particular example, I have ~300 items being returned, and resolving 5 fields (on each. The SQL Query takes about 18ms to return results, and the full HTTP response takes 78ms.

After installing your package the request takes about 18ms and full HTTP response takes 37ms. This is much more reasonable, but there still might be some opportunities for improvements. 

I ran the CPython profiler for the duration of the request, here is the breakdown of time spent in the graphql libraries with the experimental executor:

```
   ncalls  cumtime    filename:lineno(function)
        1    0.165    flask/app.py:1605(dispatch_request)
        1    0.165    flask/views.py:82(view)
        1    0.165    flask_graphql/graphqlview.py:58(dispatch_request)
        1    0.162    flask_graphql/graphqlview.py:149(execute_graphql_request)
        1    0.159    flask_graphql/graphqlview.py:146(execute)
        1    0.159    graphql/execution/executor.py:32(execute)
        1    0.159    graphql/execution/experimental/executor.py:14(execute)
        3    0.159    promise/promise.py:42(__init__)
        1    0.159    promise/promise.py:73(do_resolve)
        1    0.159    graphql/execution/experimental/executor.py:42(executor)
        1    0.159    graphql/execution/experimental/executor.py:59(execute_operation)
    323/1    0.159    graphql/execution/experimental/fragment.py:98(resolve)
   2255/1    0.155    graphql/execution/experimental/resolver.py:25(on_complete_resolver)

```

I'm using a CPython runtime in AWS, do you think your experimental executor is complete/stable enough for me to use it in production (obviously I will test it)?

**Comment by syrusakbary** [comment]
Hi @qubitron, thanks for the info and the profiling data!

I've fixed few issues in the experimental executor and now is as stable as the master branch.
For extra verification, I've executed all the `master` tests using the experimental executor and all are passing ☺️

So yes, as stable as master! :)

**Comment by mwilliamson-healx** [comment]
Unfortunately, this is still probably too slow for my use-case -- GraphJoiner is around four times faster. When profiling, it seems like most of the time is spent in (potentially asynchronous) field resolution.

Having said that, I'm not sure that the approach I'm using is really compatible with the way Graphene works. I suspect my comments aren't particularly helpful, so I'll be quiet!

**Comment by qubitron** [comment]
@mwilliamson-healx I agree it would be nice if this could be faster, for me these changes make it usable but further performance improvements would be nice. I took a cursory look at the GraphJoiner, I haven't had time to full internalize how it works and although it seems like a promising alternative, I'd prefer if the graphene approach could be made faster or if some sort of hybrid approach could be used.

One thing that would be interesting for me is if somehow we could select only the columns from SQL that were requested by the user's query, to further improve database performance.

**Comment by syrusakbary** [comment]
I'm still working on improving Performance.
First step is quite close to be ready, is a new (and ultra-performant) promise implementation.

I'm going to drop here some numbers, so is easier to see the advantages by using just the faster implementation of promise:

# Non-optimized GraphQL resolution

## Old promise
```
------------------------------------------------------------------------------------------ benchmark: 5 tests -----------------------------------------------------------------------------------------
Name (time in ms)                                      Min                 Max                Mean             StdDev              Median                IQR            Outliers(*)  Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_big_list_of_ints_serialize                     2.4519 (1.0)        4.8950 (1.0)        2.8593 (1.0)       0.4961 (1.0)        2.6586 (1.0)       0.4846 (1.0)            48;21     380           1
test_big_list_of_ints                              61.0509 (24.90)     73.8399 (15.08)     66.3891 (23.22)     3.7764 (7.61)      66.2786 (24.93)     6.3930 (13.19)            6;0      16           1
test_big_list_objecttypes_with_one_int_field      231.4451 (94.39)    274.0550 (55.99)    253.6332 (88.70)    17.2165 (34.70)    257.7021 (96.93)    27.6580 (57.08)            2;0       5           1
test_fragment_resolver_abstract                   233.4590 (95.22)    283.4949 (57.92)    259.2367 (90.66)    21.3765 (43.09)    263.5479 (99.13)    37.4374 (77.26)            2;0       5           1
test_big_list_objecttypes_with_two_int_fields     373.6482 (152.39)   407.3970 (83.23)    391.4426 (136.90)   14.5990 (29.43)    391.9201 (147.42)   26.1913 (54.05)            2;0       5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

## New promise implementation https://github.com/syrusakbary/promise/pull/23
```
------------------------------------------------------------------------------------------ benchmark: 5 tests -----------------------------------------------------------------------------------------
Name (time in ms)                                      Min                 Max                Mean             StdDev              Median                IQR            Outliers(*)  Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_big_list_of_ints_serialize                     2.4672 (1.0)        7.0231 (1.0)        2.9814 (1.0)       0.5989 (1.0)        2.7701 (1.0)       0.4563 (1.0)            40;31     378           1
test_big_list_of_ints                              23.3240 (9.45)      31.2262 (4.45)      26.8308 (9.00)      1.9695 (3.29)      26.7700 (9.66)      3.2494 (7.12)            14;0      36           1
test_fragment_resolver_abstract                   112.4361 (45.57)    160.6219 (22.87)    139.5578 (46.81)    20.4794 (34.19)    149.4532 (53.95)    35.4158 (77.61)            2;0       7           1
test_big_list_objecttypes_with_one_int_field      165.3101 (67.00)    201.4430 (28.68)    181.6540 (60.93)    15.7699 (26.33)    181.4460 (65.50)    29.1352 (63.85)            3;0       6           1
test_big_list_objecttypes_with_two_int_fields     248.4190 (100.69)   291.1139 (41.45)    267.6542 (89.77)    17.9228 (29.93)    259.4721 (93.67)    28.7293 (62.96)            2;0       5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

# Optimized GraphQL resolution

## Old Promise
```
------------------------------------------------------------------------------------------ benchmark: 5 tests -----------------------------------------------------------------------------------------
Name (time in ms)                                      Min                 Max                Mean             StdDev              Median                IQR            Outliers(*)  Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_big_list_of_ints_serialize                     2.4509 (1.0)        4.5359 (1.0)        2.9296 (1.0)       0.4356 (1.0)        2.7819 (1.0)       0.4752 (1.0)            54;25     351           1
test_big_list_of_ints                              14.3750 (5.87)      20.3481 (4.49)      16.1198 (5.50)      1.0453 (2.40)      15.9812 (5.74)      0.8274 (1.74)            15;6      65           1
test_big_list_objecttypes_with_one_int_field       73.8251 (30.12)    115.9289 (25.56)     92.0637 (31.43)    15.2907 (35.10)     82.6714 (29.72)    27.2505 (57.35)            4;0      12           1
test_big_list_objecttypes_with_two_int_fields      98.5930 (40.23)    149.9560 (33.06)    123.6130 (42.19)    19.3822 (44.50)    128.8331 (46.31)    35.7828 (75.31)            4;0       9           1
test_fragment_resolver_abstract                   115.6740 (47.20)    156.7039 (34.55)    138.5075 (47.28)    16.4670 (37.80)    146.8499 (52.79)    28.6682 (60.33)            3;0       7           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

## New Promise implementation
```
------------------------------------------------------------------------------------------ benchmark: 5 tests -----------------------------------------------------------------------------------------
Name (time in ms)                                      Min                 Max                Mean             StdDev              Median                IQR            Outliers(*)  Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_big_list_of_ints_serialize                     2.4672 (1.0)        7.0231 (1.0)        2.9814 (1.0)       0.5989 (1.0)        2.7701 (1.0)       0.4563 (1.0)            40;31     378           1
test_big_list_of_ints                              23.3240 (9.45)      31.2262 (4.45)      26.8308 (9.00)      1.9695 (3.29)      26.7700 (9.66)      3.2494 (7.12)            14;0      36           1
test_fragment_resolver_abstract                   112.4361 (45.57)    160.6219 (22.87)    139.5578 (46.81)    20.4794 (34.19)    149.4532 (53.95)    35.4158 (77.61)            2;0       7           1
test_big_list_objecttypes_with_one_int_field      165.3101 (67.00)    201.4430 (28.68)    181.6540 (60.93)    15.7699 (26.33)    181.4460 (65.50)    29.1352 (63.85)            3;0       6           1
test_big_list_objecttypes_with_two_int_fields     248.4190 (100.69)   291.1139 (41.45)    267.6542 (89.77)    17.9228 (29.93)    259.4721 (93.67)    28.7293 (62.96)            2;0       5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by syrusakbary** [comment]
When used with PyPy the difference is even bigger, and this is just the beginning.
Also, when having multiple fields in a same ObjectType, the improvement significant.

After finishing this promise implementation, I will work on separate the serializer that I assume will give another ~2x gains if using a simple `dict` instead of `OrderedDict` for serialization, and maybe even higher if serialized directly to `JSON`. This will also open the possibility of using other serializers like `msgpack` :)

And after that, optimizations with `Cython` will help to crush all benchmarks! 😊

**Comment by syrusakbary** [comment]
PS: Meanwhile I'm also working on a `dataloader` implementation for Python that will solve the **N+1** problem in GraphQL

**Comment by qubitron** [comment]
Amazing work, @syrusakbary! Looking forward to the improvements, let me know if I can help test any changes.

**Comment by qubitron** [comment]
@syrusakbary I am a bit hesitant to use PyPy, I ran into some bugs/compatibility issues with Cython libraries (unrelated to graphene) and was getting mixed performance results using sqlalchemy. That being said, if the wins are there then it's always good to have that option.

**Comment by syrusakbary** [comment]
I've been able to improve a little bit more the type resolution, about ~35%.

New benchmarks (new promise and better type resolution with experimental executor)

```
--------------------------------------------------------------------------------------- benchmark: 5 tests ---------------------------------------------------------------------------------------
Name (time in ms)                                     Min                 Max               Mean            StdDev             Median               IQR            Outliers(*)  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_big_list_of_ints_serialize                    2.6469 (1.0)        5.0581 (1.0)       2.9428 (1.0)      0.4469 (1.0)       2.7812 (1.0)      0.2511 (1.0)            47;53     401           1
test_big_list_of_ints                             13.6490 (5.16)      21.1191 (4.18)     15.1494 (5.15)     1.7030 (3.81)     14.3925 (5.18)     1.9491 (7.76)            12;2      62           1
test_big_list_objecttypes_with_one_int_field      60.2801 (22.77)     90.2431 (17.84)    67.1742 (22.83)    9.6505 (21.60)    63.0350 (22.67)    5.5089 (21.94)            2;2      15           1
test_big_list_objecttypes_with_two_int_fields     82.4349 (31.14)    110.2500 (21.80)    90.0414 (30.60)    7.7319 (17.30)    88.1380 (31.69)    9.3712 (37.32)            1;1      12           1
test_fragment_resolver_abstract                   92.1650 (34.82)    107.6009 (21.27)    98.8749 (33.60)    4.5259 (10.13)    97.8079 (35.17)    4.3540 (17.34)            2;0       8           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by syrusakbary** [comment]
(all this benchmarks are without PyPy, just plain Python with the common CPython executor)

**Comment by syrusakbary** [comment]
The latest `next-query-builder` branch now includes the ultra-performant version of promise.

Just by running `pip install pip install https://github.com/graphql-python/graphql-core/archive/features/next-query-builder.zip` it should upgrade promise to  `promise>=2.0.dev`.

(and doing `executor.use_experimental_executor = True`)

@qubitron Willing to know the performance improvements!

**Comment by qubitron** [comment]
@syrusakbary I gave this a test and I'm seeing similar performance numbers as the previous version, wish I had a different answer! Still seems to be spending about ~20ms of time in resolving in the graphql layer. 

Do you still have the previous archive available? My code has changed somewhat and will be easier for me to compare results if I can change back and forth.

**Comment by syrusakbary** [comment]
Here are the new promise 2.0+query-builder requirements:
```bash
# Installing Next query builder
pip install https://github.com/graphql-python/graphql-core/archive/features/next-query-builder.zip
# Installing promise 2.0
pip install "promise>=2.0.dev"
```

Here are the previous requirements:
```bash
# Installing Next query builder (working with old promise)
pip install https://github.com/graphql-python/graphql-core/archive/features/next-query-builder-prev.zip
# Installing promise 1.x
pip install "promise==1.0.1"
```

For verifying that the versions installed corresponds with the ones listed, you can do:
```python
import graphql
import promise
print graphql.__version__
# Only available since Promise 2.0
print promise.__version__
```

**Comment by jameswyse** [comment]
I've been trying out the experimental executor (with all libs on latest stable versions) but it seems to be slower with it enabled.

I did some basic comparison benchmarks by recording the total request time of my two largest / most complex queries with it enabled and disabled and the average results were:

- Query 1 went from 1738ms to 2142ms (404ms slower)
- Query 2 went from 1453ms -> 1749ms (296ms slower)

**Comment by syrusakbary** [comment]
Hi @jameswyse,
The experimental executor is specially suited when returning big lists of scalars or ObjectTypes with few fields. However, it should beat the normal executor in almost all benchmarks.

Could be possible to have a repo that let me reproduce it so I can optimize for it? :)

**Comment by jameswyse** [comment]
@syrusakbary that makes sense. Most of our slower queries are lists of ObjectTypes from Django models with lots of relations / deep nesting and most of the request time seems to be spent in python instantiating and resolving all these types.

It's kinda tricky to extract but I'll try to put a repo together soon 👍

**Comment by sirmarlo** [comment]
Has anyone gotten any improvement with this challenge (i.e. slow response times due to field resolution)? We are addressing this issue as well with even smaller data sets (returning 50-100 items). Our main issue is our data types are large and a bit nested so there are multiple fields that need to get resolved whenever a client has a complex query.

We've tried adding some layer of caching in the field resolution but are unable to get something feasible even if we cache the resolve functions and/or the execute call within a custom executor (that we inherit from SyncExecutor).

**Comment by syrusakbary** [comment]
Hi all!

I'm working in Quiver, the next generation GraphQL engine.
This engine works in a similar way as a high-performance template engine.

The queries will be compiled directly to python functions, so this way remove the overhead of the GraphQL framework, and the queries will be as performant as calling all the resolution functions by hand. With it we can se a 5-10x improvement over the default GraphQL engine.

Right now is closed-source and specifically directed to medium-large size companies.
So if please, you have any needs to make GraphQL as performant as possible contact me.

http://graphql-quiver.com

**Comment by syrusakbary** [comment]
Hi everyone,

Quiver is now ready to being used by the public!
I released a new article analyzing how it works:
https://medium.com/@syrusakbary/quiver-graphql-on-steroids-13612ea1ea77

**Comment by wxkin** [comment]
is there any update on this? Consider the following example, inspired by @mwilliamson-healx, where 100K `Users` need to be returned. Using graphene.Scalar the response is `x8` faster.
 

- Is there any way to drop the type checking/casting? Or anyway to speed up the resolves? 
It is not really a use case to return such number of objects however it is a use case to return a few thousand nested objects.

```python
import graphene
import cProfile
import StringIO
import pstats
from contextlib import contextmanager
from graphene.test import Client

@contextmanager
def profile(show_calls=None, message=None):
  print("\n============== Profiler start ==============" )
  if message:
    print(" " + message + " ... ")

  pr = cProfile.Profile()
  pr.enable()
  yield
  pr.disable()
  s = StringIO.StringIO()
  sortby = 'cumulative'
  ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
  if show_calls:
    ps.print_stats()
    print s.getvalue()
  else:
    print ('Execution time %f seconds' % ps.total_tt)
  print("--------------------------------------------")


class UserQuery(graphene.ObjectType):
    id = graphene.Int()

class UserAbstract(graphene.Scalar):

  @staticmethod
  def serialize(dt):
    return dt

class Query(graphene.ObjectType):
    users = graphene.Field(graphene.List(UserQuery))
    users_abstract = graphene.Field(graphene.List(UserAbstract))

    def resolve_users(self, context):
      return users

    def resolve_users_abstract(self, context):
      return resolved_users


class User(object):
    def __init__(self, id):
        self.id = id

nof_users = 100000
users = [User(index) for index in range(nof_users)]
resolved_users = [user.__dict__ for user in users]

schema = graphene.Schema(query=Query)

with profile(message="Fetch using ObjectType", show_calls=False):
  response = Client(schema).execute('{users{id}}')
  assert (len(response['data']['users']) == nof_users)


with profile(message="Fetch using Scalar", show_calls=False):
  response = Client(schema).execute('{usersAbstract}')
  assert (len(response['data']['usersAbstract']) == nof_users)
```

```
============== Profiler start ==============
 Fetch using ObjectType ... 
Execution time 8.595509 seconds
--------------------------------------------

============== Profiler start ==============
 Fetch using Scalar ... 
Execution time 1.008313 seconds
--------------------------------------------
```

**Comment by syrusakbary** [comment]
With the current execution model is almost impossible to achieve more speedup.


PS: I just tested your code using [Quiver](https://graphql-quiver.com/), and the performance Gains are considerable (10x)

### Without Quiver

```
============== Profiler start ==============
 Fetch using ObjectType ...
Execution time 3.200170 seconds
--------------------------------------------

============== Profiler start ==============
 Fetch using Scalar ...
Execution time 0.540605 seconds
--------------------------------------------
```

### With Quiver
About 10x speedup.
```
============== Profiler start ==============
 Fetch using ObjectType ...
Execution time 0.333384 seconds
--------------------------------------------

============== Profiler start ==============
 Fetch using Scalar ...
Execution time 0.048430 seconds
--------------------------------------------
```

Here is the code I used for testing:

```python
import graphene
import cProfile
from io import StringIO
import pstats
from contextlib import contextmanager

from graphql import GraphQLDeciderBackend, GraphQLCachedBackend, GraphQLCoreBackend
from graphql.backend.quiver_cloud import GraphQLQuiverCloudBackend


@contextmanager
def profile(show_calls=None, message=None):
    print("\n============== Profiler start ==============")
    if message:
        print(" " + message + " ... ")

    pr = cProfile.Profile()
    pr.enable()
    yield
    pr.disable()
    s = StringIO()
    sortby = "cumulative"
    ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
    if show_calls:
        ps.print_stats()
        print(s.getvalue())
    else:
        print("Execution time %f seconds" % ps.total_tt)
    print("--------------------------------------------")


class UserQuery(graphene.ObjectType):
    id = graphene.Int()


class UserAbstract(graphene.Scalar):
    @staticmethod
    def serialize(dt):
        return dt


class Query(graphene.ObjectType):
    users = graphene.Field(graphene.List(UserQuery))
    users_abstract = graphene.Field(graphene.List(UserAbstract))

    def resolve_users(self, context):
        return users

    def resolve_users_abstract(self, context):
        return resolved_users


class User(object):
    def __init__(self, id):
        self.id = id


# For using Quiver
QUIVER_DSN = "https://YOUR_DSN@api.graphql-quiver.com/"  # You get this DSN when registering in GraphQL-quiver.com website and creating a Project, no cost for trying it
backend = GraphQLQuiverCloudBackend(QUIVER_DSN, {"asyncFramework": None})

# For use the normal backend
# backend = GraphQLCoreBackend()


nof_users = 100000
users = [User(index) for index in range(nof_users)]
resolved_users = [user.__dict__ for user in users]

schema = graphene.Schema(query=Query)

document1 = backend.document_from_string(schema, "{users{id}}")
document2 = backend.document_from_string(schema, "{usersAbstract}")

with profile(message="Fetch using ObjectType", show_calls=False):
    response = document1.execute()
    assert len(response.data["users"]) == nof_users


with profile(message="Fetch using Scalar", show_calls=False):
    response = document2.execute()
    assert len(response.data["usersAbstract"]) == nof_users
```

**Comment by samlll42-github** [comment]
We are also having a lot of performance problem with large set (few hundreds, or few thousands) being excruciatingly slow to resolve.

Besides non-opensource options (.ie Quiver), will graphene/graphql-core ever have decent performance on large sets? Is anybody working on that or have ideas on how we could optimize? We are happy to help if there is any such initiative in progress.

**Comment by chriggi** [comment]
Same here, I wish I knew this before I implemented graphene!

**Comment by danpalmer** [comment]
We're doing something similar to this. It's a little less trivial, but not far off.

With 35k products, the following query takes around 14 seconds after we've managed to get the IDs. We're getting the IDs as just a list of integers from a cache, and then bypassing the database query that would typically generate the list if the client only requests the IDs. (The use-case here is for clients to pre-populate a full ID result set).

```graphql
query Products {
    products {
        edges {
            cursor
            node {
                id
            }
        }
    }
}
```

I realise that this is not the most optimal structure in which to return the data, but this is the format that makes most sense for the client to our API, and we'd also rather not change the architecture of parts like this for speed reasons if we can avoid it (we're obviously cool with optimising the code behind this as much as we can).

A quick test shows that the following demo code takes ~0.24s on the same machine. This is obviously missing the overheads from Django, GraphQL parsing, etc, but I think it's a reasonable lower bound.

```python
import base64
import json
import time

def item(x):
    return {
        'cursor': base64.b64encode('arrayconnection:{}'.format(x).encode('ascii')).decode('ascii'),
        'node': {
            'id': base64.b64encode('ProductType:{}'.format(x).encode('ascii')).decode('ascii'),
        }
    }

t1 = time.time()
json.dumps([item(x) for x in range(35_000)])
print(time.time() - t1)
```

I've tried this query with Quiver Cloud and I don't see any difference in the response time. I might have it set up incorrectly, but it looks ok - it's making queries to the service and receiving valid responses.

From some basic profiling, it seems that Promises are causing quite a large overhead. Of the top 10 slowest functions, totalling 4.4 seconds of run time. It also looks like there are possible optimisations in Graphene - 0.5 seconds are spent in `isinstance`, which is called 2.3 million times.

Some more advice here would be great.

Some optimisation ideas off the top of my head:

 - Graphene, Graphene-Django, GraphQL-Core, Promises, etc, all make very heavy use of functions and lambdas - these aren't particularly fast in Python when you get into the realm of detailed performance optimisation, I wonder if there's some parts that could be refactored to remove some of the function calls.
 - These libraries also make heavy use of `isinstance`, `callable`, `issubclass`, etc, to support a range of input types at various points. While this results in a nice API, and these functions aren't slow, the number of times they are being called is very significant (these 3 functions are called roughly 4.4 million times in our query above, taking ~1.3 seconds).
 - Is there a possibility for users of these libraries to return the JSON result for portions of the return data, if they are able to generate this easily? This would allow users to drop down to lower level code and manually construct responses in order to optimise if necessary.
 - Could these libraries optimise the code path for when Promises aren't being used? Anecdotally, we use Promises in ~1-2% of resolvers, and while they are really useful in these cases, it's a large penalty on other code paths - in this case roughly a 30% speed reduction.

Thoughts appreciated!

**Comment by wxkin** [comment]
We have decided to drop Graphene and use just the GraphQL-Core. At this moment we have the feeling that Graphene does not add match (we don't use all the features) but instead introduces overhead. We really don't need type checking on that level as  we can guarantee the types/attributes.

**Comment by syrusakbary** [comment]
Hi @wxkin,

I'm sorry to hear that. Graphene is just a helper on top of `graphql-core`, which improves significantly readability and ease of use of GraphQL within Python.
I've spent *a lot of time* on the overhead side and I can assure (regarding performance), that **Graphene and GraphQL-core are equally performant** (and will always be), since after Schema creation Graphene doesn't interact with the mix (ergo, no overhead on the execution side).

**Comment by syrusakbary** [comment]
@dan98765 I think all this ideas are great. Will reply inline:

Some optimisation ideas off the top of my head:
> Graphene, Graphene-Django, GraphQL-Core, Promises, etc, all make very heavy use of functions and lambdas - these aren't particularly fast in Python when you get into the realm of detailed performance optimisation, I wonder if there's some parts that could be refactored to remove some of the function calls.
Perhaps some cases can be improved. Not sure how % we can save from it (maybe 2% max?)

> These libraries also make heavy use of isinstance, callable, issubclass, etc, to support a range of input types at various points. While this results in a nice API, and these functions aren't slow, the number of times they are being called is very significant (these 3 functions are called roughly 4.4 million times in our query above, taking ~1.3 seconds).
I would very happy if we can lower the times this functions are called :)

> Is there a possibility for users of these libraries to return the JSON result for portions of the return data, if they are able to generate this easily? This would allow users to drop down to lower level code and manually construct responses in order to optimise if necessary.
You mean returning JSON without going through the GraphQL engine?

> Could these libraries optimise the code path for when Promises aren't being used? Anecdotally, we use Promises in ~1-2% of resolvers, and while they are really useful in these cases, it's a large penalty on other code paths - in this case roughly a 30% speed reduction.
That's true. Promises (or async code in general) introduce a significant overhead.

Even if only one field in a subtree (for example the `distance` field in: `getUser->photos->location->distance`) is a Promise, all parent "roots" will be wrapped in a promise for it's resolution.
That means that:
* `distance` is a promise (as is the result of executing the resolver)
* `location` will be a promise (since depends on the distance promise field)
* `photos` will be a promise (returning an array, since one or more of it's photos is a promise)
* `getUser` will be a promise

This is something that can't be changed in the current GraphQL model, given the spec.
Quiver tries to stop the Promise chain, by only using it when strictly necessary.

In the case of using just `asyncio` (without promises, with `graphql-core-next`), the same thing happens. So it's not really a matter of the library, but how the GraphQL executor is implemented.

**Comment by danpalmer** [comment]
@syrusakbary Cool, it sounds like there are fewer performance optimisations than I thought there might be.

> You mean returning JSON without going through the GraphQL engine?

Yes. In this particular instance we happen to trivially know what the JSON result would be, so a potential optimisation would be to allow the resolver to return the actual JSON that will be returned in the response. I realise this is quite a layering violation, and I don't like the idea, but it could potentially cut ~97% of the runtime of this particular query.

> Even if only one field in a subtree (for example the distance field in: getUser->photos->location->distance) is a Promise, all parent "roots" will be wrapped in a promise for it's resolution.

This makes sense, I understand that it has to go up the tree. In our case we're not returning a promise from anything in this query, so it doesn't feel worth it to wrap every resolver in a Promise layer. In this case, there are 35k nodes, each with a cursor and an ID, so I believe that would be around 105k resolutions to be made. That could be quite an overhead vs just returning the data directly.

> In the case of using just asyncio (without promises, with graphql-core-next), the same thing happens. So it's not really a matter of the library, but how the GraphQL executor is implemented.

We don't use asyncio, we're just doing synchronous execution.

I've tried the suggestion you gave me in our email:

```python
from promise import async_instance
async_instance.disable_trampoline()
```

This reduces the runtime by around half. Now 6-7 seconds to get the 35k list items, but at the cost of not being able to use promises in the few places we do make use of them.

**Comment by mjmonroe** [comment]
We use MongoDB as our database, which outputs BSON, easily converted into JSON. One MongoDB document is used to create a nested set of Graphene objects. Every Graphene object is then type checked by the GraphQl library. The type checking is a significant bottleneck. Being able to turn it off, or otherwise bypass the type checking would make our application much faster.

All the data in MongoDB is already type checked before being saved to the database via MongoEngine. In our application, type checking a second time is not necessary and just adds overhead.

**Comment by mjmonroe** [comment]
@ktosiek that sounds perfect. I would definitely use it in my graphene services since I already have a layer that validates the types, there is no need for me to have Graphene do that as well. That's a very nice speedup!

**Comment by ktosiek** [comment]
Simplified version of the monkey patch I'm using: https://gist.github.com/ktosiek/a309f772399482a47cf2c4ed219ff1af
You still have to look at `info` to know which fields should be returned, but this will skip a lot of computations.

**Comment by devop911** [comment]
I am resolving 2000entries with 2 ids, the response from ElasticSearch comes within 150ms.
Graphene needs 4 Seconds to resolve the Data.

Ist there any way for performance tweeking nowadays?

**Comment by ktosiek** [comment]
@devop911 have you tried the resolver bypass I've linked above? Other than that, I don't think there's a way - each value goes through layers of resolvers and promises. Maybe the v3 will help, I think it uses native Python async/await instead.

**Comment by qeternity** [comment]
For those of you using Django, this is a must-use: https://github.com/tfoxy/graphene-django-optimizer

Performs pretty amazing automatic select_related and prefetch_related to avoid lazy loading resolved fields

**Comment by ktosiek** [comment]
@qeternity it's really good, but I don't think this issue is about preparing the data: this is more about time spent inside graphene/graphql-core, and caling the resolvers.

Now, there is something wrong: v3 seems to be 2x slower than v2: https://gist.github.com/ktosiek/849e8c7de8852c2df1df5af8ac193287

**Comment by Miloas** [comment]
> I have a prototype that went from 0.8s to 0.3s by moving from graphene ObjectType to a dict and a simple fields filtering/renaming function.

@ktosiek  hi, thx for your awesome idea to skip expensive `complete_value` call. 

But I cant understand `moving from graphene ObjectType to a dict and a simple fields filtering/renaming function`.  

What's the meaning of moving `graphene.ObjectType` to a dict ?

It's very helpful if you give a simple example code.

**Comment by ktosiek** [comment]
@Miloas It's about using a custom function instead of Graphene resolvers. Something like:
```python
def extract_fields(info, data):
    """Extract fields requested by the client from `data`"""
    return OrderedDict(
        (field.alias.value if field.alias else field.name.value,
         data[field.name.value])
        for field in info.field_asts[0].selection_set.selections)
```

This code is just an illustration, not a tested example. The general idea is that the client might request just a selection of fields, and might even request a field under a different name (in GraphQL: `{ nameTheClientLikes: actualFieldName }`) - we don't want to lose that, even for the "raw results" path.

**Comment by qeternity** [comment]
> @qeternity it's really good, but I don't think this issue is about preparing the data: this is more about time spent inside graphene/graphql-core, and caling the resolvers.
> 
> Now, there is something wrong: v3 seems to be 2x slower than v2: https://gist.github.com/ktosiek/849e8c7de8852c2df1df5af8ac193287

Just tested locally and found the same although not nearly to the degree that you did (v3 is ~25-50% slower)

`(env3)  ~/Code/graphene  head -n 50 cprof.txt
53.549506894000004
         99251786 function calls (93244536 primitive calls) in 52.909 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
36016108/36016098    8.114    0.000   12.218    0.000 {built-in method builtins.isinstance}
2000010/10    6.241    0.000   53.415    5.342 execute.py:701(complete_value)
  7000070    6.081    0.000   15.722    0.000 inspect.py:221(isawaitable)
1000010/10    2.889    0.000   53.416    5.342 execute.py:570(resolve_field)
1000010/10    2.537    0.000   53.416    5.342 execute.py:416(execute_fields)
2000010/10    2.340    0.000   53.415    5.342 execute.py:640(complete_value_catching_error)
  7000072    2.291    0.000    2.291    0.000 {built-in method _abc._abc_instancecheck}
       10    1.915    0.192   53.415    5.342 execute.py:785(complete_list_value)
  7000072    1.813    0.000    4.104    0.000 abc.py:137(__instancecheck__)
  1000010    1.684    0.000    7.965    0.000 execute.py:607(resolve_field_value_or_error)
  1000010    1.361    0.000    2.113    0.000 execute.py:547(build_resolve_info)
  1000010    1.245    0.000    1.577    0.000 values.py:141(get_argument_values)
  1000000    1.239    0.000    2.113    0.000 resolver.py:9(dict_or_attr_resolver)
  1000000    1.165    0.000   34.578    0.000 execute.py:974(collect_and_execute_subfields)
  3000785    0.930    0.000    0.930    0.000 {built-in method __new__ of type object at 0x10cb3ce10}
  1000000    0.904    0.000    2.226    0.000 execute.py:839(complete_leaf_value)
  2000010    0.865    0.000    1.163    0.000 is_nullish.py:9(is_nullish)
  1000000    0.796    0.000    1.086    0.000 scalars.py:35(serialize_int)
  1000000    0.791    0.000    1.315    0.000 execute.py:985(collect_subfields)
  1000000    0.683    0.000    1.367    0.000 path.py:14(add_key)
  1000000    0.661    0.000   35.239    0.000 execute.py:939(complete_object_value)
  2000040    0.651    0.000    1.038    0.000 definition.py:1449(is_non_null_type)
  2000020    0.650    0.000    1.043    0.000 definition.py:1575(is_leaf_type)
  2000010    0.615    0.000    0.945    0.000 definition.py:1404(is_list_type)
  1000010    0.582    0.000    0.776    0.000 execute.py:1040(get_field_def)
  2001433    0.517    0.000    0.520    0.000 {method 'get' of 'dict' objects}
  1000000    0.378    0.000    0.665    0.000 resolver.py:1(attr_resolver)
  1000000    0.305    0.000    0.533    0.000 definition.py:1609(is_abstract_type)
  1000133    0.301    0.000    0.441    0.000 definition.py:773(is_object_type)
  2000450    0.301    0.000    0.301    0.000 typing.py:893(cast)
  1014310    0.295    0.000    0.295    0.000 {built-in method builtins.getattr}
  2000210    0.272    0.000    0.272    0.000 {method 'items' of 'dict' objects}
  1000000    0.236    0.000    0.236    0.000 is_invalid.py:8(is_invalid)
  1003021    0.214    0.000    0.214    0.000 {method 'append' of 'list' objects}
  1000000    0.201    0.000    0.201    0.000 {built-in method builtins.id}
  1000010    0.200    0.000    0.200    0.000 values.py:152(<dictcomp>)
      244    0.164    0.001    0.164    0.001 {method 'read' of '_io.FileIO' objects}
        1    0.108    0.108   53.550   53.550 <timeit-src>:2(inner)
        1    0.060    0.060    0.101    0.101 test.py:16(<listcomp>)
   100000    0.041    0.000    0.041    0.000 test.py:13(__init__)
      244    0.029    0.000    0.029    0.000 {built-in method marshal.loads}
       14    0.025    0.002    0.029    0.002 {built-in method _imp.create_dynamic}
      244    0.015    0.000    0.179    0.001 <frozen importlib._bootstrap_external>:914(get_data)
     1038    0.015    0.000    0.015    0.000 {built-in method posix.stat}`

**Comment by wojcikstefan** [comment]
The example given in this issue's description is outdated (unless I'm missing something). @syrusakbary does it make sense to close this issue since it's misleading to newcomers (like myself)? A separate issue could be open about further performance improvements.

---

Context: The original issue says:

> the example below returns 10000 objects with an ID field, and that takes around ten seconds to run.

This is no longer true. Here's the setup code that, AFAICT, is the modern equivalent of the code from the issue's description:
```py
import graphene

class User(object):
    def __init__(self, id):
        self.id = id

users = [User(index) for index in range(0, 10000)]

class UserQuery(graphene.ObjectType):
    id = graphene.Int()

class Query(graphene.ObjectType):
    users = graphene.List(UserQuery)

    def resolve_users(root, info):
        return users

schema = graphene.Schema(query=Query)
```

Now, this consistently gives ~300-400ms on my 2018 MBP, not 10s as the issue description suggests. Example:
```
In [28]: %timeit rv = schema.execute('{ users { id } }')
1 loop, best of 3: 336 ms per loop
```

**Comment by mwilliamson** [comment]
Although the example is faster than when I originally opened the issue, it's still too slow for my use-case, and is around 10-15 times slower than the library we're currently using.

**Comment by wojcikstefan** [comment]
Yep, there's definitely plenty of room for improvement still. I just argue that this issue has become outdated, misleading, and quite messy overall :) Would be good to start afresh with a new issue mentioning the accurate status quo and what next steps can be taken.

**Comment by Anon731** [comment]
I made following snippet which allows you to use custom connection class with improved performance: https://gist.github.com/Anon731/eb7ace0bd596bbd45496e3cd12f2eafd

Performance comparison:

3k Candle objects in db.

Regular Edges query 2.5 - 3 seconds approximately (20 iterations):
```graphql
{
  candleList(closingTime_Gt: "2019-09-24T00:00:00+00:00") {
    edges {
      node {
        id
        createdAt
        openingPrice
        lowPrice
        highPrice
        closingPrice
      }
    }
  }
}
```

ScalarEdges Query 450ms - 600ms approximately (20 iterations):
```graphql
query {
  candleList(closingTime_Gt: "2019-09-24T00:00:00+00:00") {
    scalarEdges(fields: [
      "id",
      "created_at",
      "opening_price",
      "low_price",
      "high_price",
      "closing_price",
    ]) {
      scalarNode
    } 
  }
}
```

If you improve serialization you can achieve x2 speedup boost.

**Comment by jedie** [comment]
I also make some tests and compare `graphene-django` with `django-rest-framework`.

In both cases i used a view with pagination (`DjangoConnectionField` or `PageNumberPagination`).
The tests requests 100 items from existing 1000 items.


**graphene:**
```
django v1.11.26
graphene v2.1.8
graphene-django v2.7.1

timeit... use 5 * 75 loop...
max...: 10.84 ms
median: 10.66 ms
min...: 10.61 ms
cProfile stats for one request: 30148 function calls (28688 primitive calls) in 0.019 seconds
```

**Rest-API:**
```
django v1.11.26
Rest-Framework v3.9.4

timeit... use 5 * 213 loop...
max...: 3.81 ms
median: 3.63 ms
min...: 3.58 ms
cProfile stats for one request: 14171 function calls (13142 primitive calls) in 0.007 seconds
```
What immediately stands out is the much higher number of function calls.

I also run a test with graphene v3 with django 2.2 with the graphene-django sources from: https://github.com/graphql-python/graphene-django/pull/812 It's ~30% slower than graphene v2 
I hope this is slower because it's not the final code yet.

**Comment by jkimbo** [comment]
@jedie could you share the code you used to benchmark Graphene vs Rest Framework?

**Comment by jedie** [comment]
> @jedie could you share the code you used to benchmark Graphene vs Rest Framework?

Sorry, can't share the code. But it's really a minimal example code.

But i made another tests and benchmark only with graphql-core: https://gist.github.com/jedie/581444e02e784ff7c2b9fb1e763759fa

It fetches only a list of 1000 dummy items and takes ~20ms

**Comment by jedie** [comment]
Now, i also made a similar test with tartiflette.

To my surprise: **tartiflette** (~57ms) is significantly slower than **graphql-core** (~20ms).

My benchmark code:

**tartiflette**: https://gist.github.com/jedie/45ddf8ee7e24704c9485eb8cbcf9ba13
**graphql-core**: https://gist.github.com/jedie/581444e02e784ff7c2b9fb1e763759fa

**Comment by sostholm** [comment]
I was fixing a performance issue in graphene-mongo:
https://github.com/graphql-python/graphene-mongo/issues/125

My pull request brought down the response time from 2s to 0.02s on a dataset of 12000 documents in MongoDB.

The solution was to provide the list_slice_length in default_resolver to prevent the default resolver from doing a len() on the collection.

It would appear that the default behavior for many orms when doing len on their collections is to load all objects in the collection.

Although I resolved this particular issue, there were plenty more like this one. I stopped trying because it would require some major changes to Graphene in order to fix it.

Will issues like this be fixed for v3?

**Comment by dodafin** [comment]
> @qeternity it's really good, but I don't think this issue is about preparing the data: this is more about time spent inside graphene/graphql-core, and caling the resolvers.
> 
> Now, there is something wrong: v3 seems to be 2x slower than v2: https://gist.github.com/ktosiek/849e8c7de8852c2df1df5af8ac193287

There doesn't seem to be this dicrepancy between 2.1.8 and 3.07b

Graphene 2: 12.702938017901033
Graphene 3: 12.651812066091225

**Comment by ktosiek** [comment]
> There doesn't seem to be this dicrepancy between 2.1.8 and 3.07b
> 
> Graphene 2: 12.702938017901033
> Graphene 3: 12.651812066091225

That particular problem was fixed in https://github.com/graphql-python/graphql-core/issues/54, fix was released in graphql-core 3.1.0.

**Comment by cancan101** [comment]
FWIW, I am using `graphql-core==3.1.4` with Ariadne and am still seeing a fair bit of unexplained time spent when returning larger data sets. This is at least one place that I am seeing a lot of time being spent (in various forms of completion vs resolving of values):
```
Function: resolve_field at line 578

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   578                                               def resolve_field(
   579                                                   self,
   580                                                   parent_type: GraphQLObjectType,
   581                                                   source: Any,
   582                                                   field_nodes: List[FieldNode],
   583                                                   path: Path,
   584                                               ) -> AwaitableOrValue[Any]:
   585                                                   """Resolve the field on the given source object.
   586                                           
   587                                                   In particular, this figures out the value that the field returns by calling its
   588                                                   resolve function, then calls complete_value to await coroutine objects,
   589                                                   serialize scalars, or execute the sub-selection-set for objects.
   590                                                   """
   591      2189       1355.0      0.6      1.3          field_node = field_nodes[0]
   592      2189       1523.0      0.7      1.5          field_name = field_node.name.value
   593                                           
   594      2189       4577.0      2.1      4.4          field_def = get_field_def(self.schema, parent_type, field_name)
   595      2189       1159.0      0.5      1.1          if not field_def:
   596                                                       return Undefined
   597                                           
   598      2189       1265.0      0.6      1.2          resolve_fn = field_def.resolve or self.field_resolver
   599                                           
   600      2189       1128.0      0.5      1.1          if self.middleware_manager:
   601      2189       4540.0      2.1      4.4              resolve_fn = self.middleware_manager.get_field_resolver(resolve_fn)
   602                                           
   603      2189      10609.0      4.8     10.3          info = self.build_resolve_info(field_def, field_nodes, parent_type, path)
   604                                           
   605                                                   # Get the resolve function, regardless of if its result is normal or abrupt
   606                                                   # (error).
   607      4378      25759.0      5.9     25.0          result = self.resolve_field_value_or_error(
   608      2189       1140.0      0.5      1.1              field_def, field_nodes, resolve_fn, source, info
   609                                                   )
   610                                           
   611      4378      48599.0     11.1     47.2          return self.complete_value_catching_error(
   612      2189       1314.0      0.6      1.3              field_def.type, field_nodes, info, path, result
   613                                                   )
```

**Comment by kevindice** [comment]
In case this helps anyone, I ran some benchmarks against graphene-django and DRF with django-silk and discovered the `FieldTracker` from `django_model_utils` was the cause of my performance issues.  The profile showed a heinous amount of time spent on the `deepcopy` function.

To paint a picture here:

```
from model_utils import FieldTracker

class Profile(models.Model):
    bio = models.TextField(blank=True)
    bio_hashtags = models.ManyToManyField(Hashtag, blank=True)

    tracker = FieldTracker()

    def save(self, *args, **kwargs):
        if self.id and self.tracker.has_changed('bio'):
            self.reconcile_bio_hashtags()
        super.save(*args, **kwargs)
```

Fetching 50 profiles, the following timings were obtained:
- DRF w/ field-tracker enabled: 10.2 seconds
- graphene-django (using relay fwiw): 9.7 seconds
- DRF w/ field-tracker disabled: 0.159 seconds
- graphene-django w/ field-tracker disabled: 0.7 seconds


0.7 seconds is still pretty bad for a query of 50 things and a single postgres query, but 90% of my problem was not graphene.

**Comment by Kobold** [comment]
Great find @kevindice . Thank you for posting!

**Comment by tomduncalf** [comment]
Just to offer my experience, it seems like performance is still an issue. Returning a set of 50 items (I'd consider them medium sized maybe? Nothing unusual), requests were taking over 2 seconds to return on a Heroku free dyno – nearly all of this time is spent in GraphQL code according to NewRelic, I had optimised the queries already. Switching to DRF greatly improved performance, so it definitely seems like GraphQL is the issue. 

I would have liked to investigate more but I am on a deadline, so I ended up switching to Rails with graphql-ruby which, to my surprise, is faster than either Python solution – the same query on Heroku with Rails returns in 300-400ms, so it's several times faster and makes the difference between a good and bad user experience. 

I prefer the developer ergonomics of Python and Django but to be honest Django and Rails are similar enough in many ways that it's not a big deal for me to switch. Obviously you can't do this if you're deep into a project, but for anyone considering Python for a GraphQL project I think it's worth being aware of these potential performance issues.

**Comment by jkimbo** [comment]
@tomduncalf were you using Graphene v2 or v3? Also I'm surprised that switching to Strawberry didn't help. Based on these benchmarks: https://twitter.com/jayden_windle/status/1235323199220592644 I would expect Strawberry or Graphene v3 to be significantly better especially for lists of objects. 

IMO I would expect GraphQL to always have a bit of a performance overhead compared to a REST endpoint since it's doing quite a bit more. It would be good to get Python performance to a point where it's comparable to other similar languages though (like ruby). Can you share any of your code?

**Comment by tomduncalf** [comment]
@jkimbo This was on Graphene v3. Unfortunately I can't share the code as it's not open source and I don't really have the time to dig into exactly why it was slower right now, but if I do find time to make a simple repro comparing Python vs Ruby I will post it here!

**Comment by jkimbo** [comment]
So @tomduncalf you properly nerd sniped me with this and I ended up building a more "realistic" benchmark. I implemented an API to fetch the top 250 rated IMDB movies in Graphene, Strawberry, DjangoRestFramework and a plain JSON api, all hosted on Django. All the data comes from a sqlite db. The code is here: https://github.com/jkimbo/django-graphql-benchmarks and you can try it out the Graphene API [here.](https://django-graphql-benchmark.herokuapp.com/graphene-graphql/#query=%7B%0A%20%20top250%20%7B%0A%20%20%20%20id%0A%20%20%20%20imdbId%0A%20%20%20%20title%0A%20%20%20%20year%0A%20%20%20%20imageUrl%0A%20%20%20%20imdbRating%0A%20%20%20%20imdbRatingCount%0A%20%20%20%20director%20%7B%0A%20%20%20%20%20%20id%0A%20%20%20%20%20%20name%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D)

Here are the P99 results against the Heroku instance:

![requests-time](https://user-images.githubusercontent.com/691952/130456503-03985e43-2fb0-4826-871f-f44f8f4f85a4.png)
![requests-pre-second](https://user-images.githubusercontent.com/691952/130456915-49816957-6a2a-4d3c-abbc-884b22cf61e3.png)

So you can see that Graphene (v3) and Strawberry are pretty much neck and neck (which is what I would expect considering that they are just different ways to setup a graphql-core server). DRF is definitely faster (~25% faster) and the the plain json endpoint is faster still. I couldn't replicate the 2 second response times you were seeing with your API @tomduncalf so I'm not sure what is going on there.

Overall GraphQL in Python is definitely slower that using something like DjangoRestFramework but not horribly so in my opinion. There are definitely things that can be improved though and thanks to this exercise I have some ideas for improvements we can make to Strawberry.

Would be interested in how this all compares to graphql-ruby as well but unfortunately my experience there is lacking.

**Comment by tomduncalf** [comment]
Hey @jkimbo, thanks for doing this and I hope my initial post didn't come across too negatively – I just wanted to share my experience for anyone else in my situation (i.e not familiar with either Django or Rails and looking to pick one for a GraphQL project), as I didn't really find much online comparing the two for GraphQL specifically and I didn't realise there is a performance overhead.

You prompted me to do a little bit more digging as I felt bad for just saying it was slow 🙃 as your demo seems to perform pretty well. One thing I didn't think to mention is that I am using Relay with my API – I did a little bit of testing with my API and it seems like using Relay adds a fairly significant overhead – almost doubling the response times on Heroku for the same query vs. a non-Relay version! I wonder if you could try using Relay with Graphene on yours and see if you see similar results? Or if you tell me how to run yours, I can try it (pretty new to Python so couldn't work out how to run yours from a git clone).

My API does still seem quite slow compared to yours and I'm not really sure why, as yours is returning a larger set of data. I'm new to Django so I could be doing something a bit stupid somewhere. To be honest, I am going to stick with Rails at this point as it's probably a slightly better fit for what I am trying to do (build an API with as little code as possible basically, haha – the ecosystem of gems seems a bit more developed for some of the things I want to do), but if you have any suggestions of good ways to profile my Python I could give it a go.

Anyway, you piqued my curiosity so I reproduced your demo in Rails! The code is at https://github.com/tomduncalf/rails-graphql-benchmark and I've deployed it to Heroku in the [EU region](https://rails-graphql-benchmark-eu.herokuapp.com/graphiql). It seems like it returns a bit faster than yours, but not dramatically so – I'm not sure how you run your benchmark but I'd be happy to try it on mine if it's useful for comparison.

There are two queries you can run, one Relay and one non-Relay (doesn't seem the Ruby version of GraphiQL supports embedding them in the URL!):

```
{
  movies {
    edges {
      node {
        id
        imdbId
        title
        year
        imageUrl
        imdbRating
        imdbRatingCount
        director {
          id
          name
        }
      }
    }
  }
}

{
  moviesPlain {
    id
    imdbId
    title
    year
    imageUrl
    imdbRating
    imdbRatingCount
    director {
      id
      name
    }
  }
}
```

Cheers,
Tom

**Comment by Ashish-Bansal** [comment]
Last year I spent a few weekends pulling off a very minimal PoC based on @syrusakbary idea of generating template code to improve GQL performance in python. 

Here's the link to the same - https://github.com/Ashish-Bansal/graphql-jit

It's very vague, un-tested PoC implementation, and needs a lot of work. 

I won't get time to work on it, so in case anyone is interested, they can work over that PoC.

**Comment by cglacet** [comment]
This thread is interesting but kind of hard to get a grasp on. That would be great if someone competent, like a maintainer of `graphql-core` or `graphene` could make a summary of the different solutions that still exist today (potential gain depending on the use case, potential issues, required version of `graphene`/`graphql-core`, maturity of the solution). 

I also have a few questions:

* Is there a significant performance benefit in moving from v2 to v3? 
* Are there significant performance differences amongst python versions?
* Can working with `graphene-sqlalchemy` be an issue when it comes to performances?
* What tool(s) are you using to measure asynchronous code performances with decent precision and without interferences? (I tried [yappi](https://github.com/sumerc/yappi) and it seems decently close to clock time but the output is not very readable). I first started to use `cProfile` because I had no idead it couldn't measure async code.
*  What is considered as a "large" dataset and how does size impact performances? I feel jus like @jkimbo and his benchmark look quite "realistic" to me (strangely when I experiment with the Graphene API I get response time under 200ms).

In my case, all requests I try to make take an enormous time to resolve (300ms for < 10 fields queries, and up to 3s for larger queries). On the other hand I sometimes get quite lower response time (~ x3 improvement) for the exact same query on the exact same database (same data state), all of which use the exact same docker image, sometimes I wonder if it's not a lack of memory/CPU issue which is causing this.

Anyway, thanks for this discussion, I hope someone smarter than me would be able to summarize the good ideas that lies here and there in this thread.

**Comment by flbraun** [comment]
@cglacet 
>Is there a significant performance benefit in moving from v2 to v3?

Currently asking myself the same question. I took the liberty to fork jkimbo's benchmark suite mentioned above, update the frameworks to their latest version and include graphene v2 as well: [flbraun/django-graphql-benchmarks](https://github.com/flbraun/django-graphql-benchmarks)

Here's the result from a fresh bench I did this morning:
![results](https://github.com/flbraun/django-graphql-benchmarks/raw/268ed121723825c9068e12be6f6d0de8c7f15e0a/latest.png)

As you can see the difference between v2 and v3 is pretty much non-existent, _however_ the benchmark suite currently only serializes a bunch of ObjectTypes. Since this is the most primitive building block of a GraphQL API and doesn't cover real-world setups utilizing pagination, connections, etc.

Maybe this is of any interest for you.

**Comment by erikwrede** [comment]
@flbraun nice update! Probably makes sense to add some asyncio into the mix as this adds a lot of overhead as well.

**Comment by flbraun** [comment]
I updated my forked bench suite to be tested against multiple Python versions, see [results here](https://github.com/flbraun/django-graphql-benchmarks#latest-results).   

tl;dw: Graphene (both v2 and v3) perform almost identical on the same Python version. However, Graphene seems to have heavily benefited from performance improvements in the Python interpreter. Jumping from 3.10 to 3.11 alone shaves off **~20%**  off of mean response times, which is kinda impressive.  
Maybe somebody with more knowledge about the inner workings of Graphene (and graphql-core) can leverage this information.

**Comment by qeternity** [comment]
If you're running this in prod, you should try pyston...we've seen perf that is close to pypy but without warmup/other issues

**Comment by pfcodes** [comment]
> Hi.
> 
> I'm curious what the latest is on this? Were noticing similar performance issues as OP posted, where a lot of time appears to be spent planning SQL Queries
> 
> Our SQL takes 60-100ms to process, but time to first byte is close to 1 second. Our profile implies a lot of time is spent planning SQL
> 
> Our Stack:
> 
> * Python 3.11
> * Django 4.2.8
> * DRF 3.14.0
> * Django Graphene 3.1.5
> 
> All optimized SQL, and we get numbers like:
> 
> Chrome Dev Tools
> 
> * 6 SQL Querues 56.44 ms
> * 44us to send request
> * 937 ms waiting for server response
> * 8.48 ms to download
> * Total ~1 second for a 100ms DB query!
> 
> Our PySpy on a Postgres and Django all running local implies much of the time is in Graphene / SQL planning?
> 
> ![profile-gunicorn-with-signed-url](https://private-user-images.githubusercontent.com/65011/289179874-8cc4f290-b69b-47c2-83e4-c30f1213db6e.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAyNDI2ODgsIm5iZiI6MTcyMDI0MjM4OCwicGF0aCI6Ii82NTAxMS8yODkxNzk4NzQtOGNjNGYyOTAtYjY5Yi00N2MyLTgzZTQtYzMwZjEyMTNkYjZlLnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA2VDA1MDYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhZDU5MmJlNmNlODYxNDNiMGQ4YjhjYjE2ODFmNjRiMzE2NTQzMzhmZmZlYWY5Yjk3ODZlMzUzNzI1YjkwYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TfrsNxlbRpQunfw8gAIcvQ4d4pMtPVUlRlrU8-3KppI)

Did you ever figure this out? @vade

**Comment by vade** [comment]
We've done a lot of optimization on our stack in the interim, so I don't know if I can speak specifically to any *one* single change, but thing we've observed. 

We use Relay, and pagination

* Using an ORM SQL optimizer makes a world of difference, especially ones that pay close attention to detail. We've been helping debug [graphene-django-query-optimizer](https://github.com/MrThearMan/graphene-django-query-optimizer) which supports only optimization, fetch related and selected related, but most importantly it optimizes nested query sets and field objects and supports filtering and relay pagination. 
* We optimized the shit out of our GQL queries and redesigned our schema so things can be a bit more light weight in terms of fields we fetch
* For some filtered query sets,  resolving total count (for relay pagination) is highly suboptimial and total count takes a majority of the SQL time and for some reason the total count breaks some optimization paths. This one of our last large hurdles. 

We've seen 10x improvement in response time with some of the above. 

I know it's hand wavy, but a lot of it was just really pay attention to details and ensuring that more complicated queries do in fact get optimized. 

We do some hand tuned query set tweaks in some cases for fields that require model / db lookups and annotate them, and we found some hot spots where we unintentionally did dumb shit like evaluate a query set in place rather than use annotated values so the DB could do the work.

cc @rsomani95 - any thing else to add from the work we did that maybe im missing?

**Comment by dicknetherlands** [comment]
For the very specific use-case that I'm working with, which is synchronous-only and does not support the full range of GraphQL features, I've made a Gist which shows how I worked around this performance issue to get an apparent speedup of 10x in some cases (not formally measured). If your use-case is similar to mine then it might make a useful starting point. https://gist.github.com/dicknetherlands/2f6e8619409fa155a05b3a863f10269a

It works by assuming that the developers will respect the schema. The key speedups come from discovering the type of each field only once, bypassing serializers for builtin scalars, bypassing nullability checks, not doing object type checks for nested fields, not re-validating the schema upon every query, and more efficient resolution of SyncFuture objects returned by graphql_sync_dataloader (if you're using that library).
